File: data-structures/linked-list/linked-list-class/java/LinkedListv1.java
import java.util.NoSuchElementException;

/**
 * LinkedListv1 is a class that represents a singly linked list.
 * It allows insertion at the front or the back of the linked list,
 * deletion at the front or the back of the linked list,
 * searching for nodes, and getting the index of a given node.
 */
public class LinkedListv1 {

    private Node head;
    private Node tail;
    private int size;

    /**
     * Constructs an empty LinkedListv1.
     */
    public LinkedListv1() {
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    /**
     * Inserts a new node at the front of the LinkedListv1.
     *
     * @param value the value to be inserted
     */
    public void pushFront(String value) {
        Node node = new Node(value, null);
        if (this.head == null) {
            this.head = node;
            this.tail = node;
        } else {
            node.setNext(this.head);
            this.head = node;
        }
        this.size += 1;
    }

    /**
     * Inserts a new node at the end of the LinkedListv1.
     *
     * @param value the value to be inserted
     */
    public void pushBack(String value) {
        Node node = new Node(value, null);
        if (this.head == null) {
            this.head = node;
            this.tail = node;
        } else {
            this.tail.setNext(node);
            this.tail = node;
        }
        this.size += 1;
    }

    /**
     * Deletes the node at the front of the LinkedListv1.
     */
    public void deleteFront() {
        if (this.head == null)
            throw new NoSuchElementException();
        if (this.head == this.tail) {
            this.head = null;
            this.tail = null;
        } else {
            this.head = this.head.getNext();
        }
        this.size -= 1;
    }

    /**
     * Deletes the node at the end of the LinkedListv1.
     */
    public void deleteBack() {
        if (this.head == null)
            throw new NoSuchElementException();
        if (this.head == this.tail) {
            this.head = null;
            this.tail = null;
        } else {
            Node node = this.head;
            while (node.getNext() != this.tail) {
                node = node.getNext();
            }

            this.tail = node;
            node.setNext(null);
        }
        this.size -= 1;
    }

    /**
     * Checks if the LinkedListv1 contains a specific value.
     *
     * @param value the value to search for
     * @return true if the LinkedListv1 contains the value, false otherwise
     */
    public boolean contains(String value) {
        if (this.head == null)
            throw new NoSuchElementException();
        Node node = this.head;
        while (node != null) {
            if (node.getValue().equals(value)) {
                return true;
            }
            node = node.getNext();
        }

        return false;
        // return indexOf(value) != -1
    }

    /**
     * Returns the index of the first occurrence of a value in the LinkedListv1.
     *
     * @param value the value to search for
     * @return the index of the value if found, -1 otherwise
     */
    public int indexOf(String value) {
        if (this.head == null)
            throw new NoSuchElementException();
        Node node = this.head;
        int i = 0;
        while (node != null) {
            if (node.getValue().equals(value)) {
                return i;
            }
            node = node.getNext();
            i += 1;
        }
        return -1;
    }

    /**
     * Returns the size of the LinkedList.
     *
     * @return The size of the LinkedList.
     */
    public int getSize() {
        return this.size;
    }

}

File: data-structures/linked-list/linked-list-class/python/linked_list_v1.py
from node import Node


class LinkedListv1:
    """
    A class representing a singly linked list.

    Attributes:
        _head: The head node of the linked list.
        _tail: The tail node of the linked list.
        _size: The size of the linked list.

    Methods:
        __init__(): Initializes an empty linked list.
        get_size() -> int: Returns the size of the linked list.
        push_front(value: str) -> None: Inserts a node with the given value at the front of the linked list.
        push_back(value: str) -> None: Inserts a node with the given value at the end of the linked list.
        delete_front() -> None: Deletes the node at the front of the linked list.
        delete_back() -> None: Deletes the node at the end of the linked list.
        contains(value: str) -> bool: Checks if the linked list contains a node with the given value.
        indexOf(value: str) -> int: Returns the index of the first occurrence of a node with the given value.
    """

    def __init__(self):
        """
        Initializes an empty linked list.
        """
        self._head = None
        self._tail = None
        self._size = 0

    def get_size(self) -> int:
        """
        Returns the size of the linked list.

        Returns:
            The size of the linked list as an integer.
        """
        return self._size

    def push_front(self, value: str) -> None:
        """
        Inserts a node with the given value at the front of the linked list.

        Args:
            value: The value to be inserted in the new node.

        Returns:
            None
        """
        node = Node(value, None)
        if self._head == None:
            self._head = node
            self._tail = node
            self._size += 1
            return

        node.set_next(self._head)
        self._head = node
        self._size += 1

    def push_back(self, value: str) -> None:
        """
        Inserts a node with the given value at the end of the linked list.

        Args:
            value: The value to be inserted in the new node.

        Returns:
            None
        """
        node = Node(value, None)
        if self._head == None:
            self._head = node
            self._tail = node
            self._size += 1
            return

        self._tail.set_next(node)
        self._tail = node
        self._size += 1

    def delete_front(self) -> None:
        """
        Deletes the node at the front of the linked list.

        Returns:
            None

        Raises:
            ValueError: If the linked list is empty.
        """
        if self._head == None:
            raise ValueError('Linked list empty')

        if self._head == self._tail:
            self._head = None
            self._tail = None
            self._size -= 1
            return

        self._head = self._head.get_next()
        self._size -= 1

    def delete_back(self) -> None:
        """
        Deletes the node at the end of the linked list.

        Returns:
            None

        Raises:
            ValueError: If the linked list is empty.
        """
        if self._head == None:
            raise ValueError('Linked list empty')

        if self._head == self._tail:
            self._head = None
            self._tail = None
            self._size -= 1
            return

        node = self._head
        while node.get_next() != self._tail:
            node = node.get_next()

        node.set_next(None)
        self._tail = node
        self._size -= 1

    def contains(self, value: str) -> bool:
        """
        Checks if the linked list contains a node with the given value.

        Args:
            value: The value to search for.

        Returns:
            True if a node with the given value is found, False otherwise.

        Raises:
            ValueError: If the linked list is empty.
        """
        if self._head == None:
            raise ValueError('Linked list is empty')
        node = self._head
        while node != None:
            if value == node.get_value():
                return True
            node = node.get_next()
        return False

    def indexOf(self, value: str) -> int:
        """
        Returns the index of the first occurrence of a node with the given value.

        Args:
            value: The value to search for.

        Returns:
            The index of the node with the given value as an integer. Returns -1 if the value is not found.

        Raises:
            ValueError: If the linked list is empty.
        """
        if self._head == None:
            raise ValueError('Linked list empty')

        node = self._head
        i = 0
        while node != None:
            if node.get_value() == value:
                return i
            node = node.get_next()
            i += 1

        return -1
